### ISSUE TITLE
Multiple Java files compiled to one Javascript file 

### ISSUE THREAD

**Issue by panzupa** [cincheo/jsweet#160]
Hi all, 

First of all I would like to thank you for your work on this excellent tool / framework. I'm using it for some time and it's great improvement to my work. 

I'd like to ask you to add functionality to jsweet that would allow me to generate one JS file from multiple Java source files. Even support for just two source files would be great for me. 

Currently I've got many java files that reuse the same code and it is duplicated among those files. Would be great to have it only one place. 

Do you think it's possible?

**Comment by lgrignon** [comment]
it shouldn't be too complicated since it is a tsc option

**Comment by lgrignon** [comment]
it shouldn't be too complicated since it is a tsc option. Did you try the bundle option? If you use the maven plugin: https://github.com/lgrignon/jsweet-maven-plugin

Even without this option, you should be able to share your code between files without duplicating it, you just have to include generated js files in the right order.

Did I miss something?

**Comment by panzupa** [comment]
Hi Igrignon, 

I just tried doing that with maven plugin but I get a lot of errors from my own 'candy'.  I have some java classes with @Ambient annotation and each of them is not found. 
ERROR can't find namespace 'xxx' ... 

Maybe I should have prepared some ts files for my candy but I don't know how...

**Comment by lgrignon** [comment]
It depends on what you want to do. Given what you say, I guess that you want to use an existing Javascript library / source code from a JSweet program?
If so, you could indeed annotate your definitions with @Ambient but it requires you to provide the ts definition files.

One technique I like is to create the javascript source code definition with interfaces (@Interface abstract class XXX should do the trick) and then cast the required object to your interface. However it is not good if you have several entry points.

ex:
```
@Interface
abstract class LibEntryPoint {
  void startSomething();
}

LibEntryPoint myLibEntryPoint = (LibEntryPoint ) window.$get("myLib");
myLibEntryPoint.startSomething();
```

or you could just translate your javascript code to JSweet code :)

**Comment by renaudpawlak** [comment]
@panzupa maybe you have hit a JSweet bug...
Normally, when you annotate with @Ambient, classes are generated as declarations (you can check your generated TypeScript code). So you don't need to provide any TypeScript definitions. However, if you say that the same code works without bundling, and then fails compiling when you use the bundle option, then I suspect a bug...
Could you check the generated bundle and tell us if the declarations are still generated?
I will also check @Ambient behavior with the bundle option ASAP.
Thanks

**Comment by renaudpawlak** [comment]
@lgrignon normally, when you use @Ambient, you don't need to provide definition files

**Comment by lgrignon** [comment]
woops, my bad, it's been a long time :)

**Comment by panzupa** [comment]
Exactly, I don't have d.ts file for my candy. What I will try to do this evening is to generate such file and pack my candy to separate jar file (right now it's just a package within bigger project. Javascript generated is correct though :-) 

I'm using eclipse plugin and just noticed that I can select in properties to generate d.ts files, however files are generated for each java source file not for whole package def.<my_candy> 

Btw when I will have this one d.ts file for my candy  will eclipse stop showing warnings in editor 'cannot find name <class from my candy >?

**Comment by renaudpawlak** [comment]
Umm. Your use case is still unclear to me actually.

Are you using an external JS lib that you are mapping with @Ambient declarations? In that case, why not using one of the JSweet pre-packaged candies?

Or are you trying to publish a JSweet program to use it as a candy from another JSweet program? In that case, why are you using @Ambient? (no need)

Can you clarify your use case first? :)

**Comment by panzupa** [comment]
Hi renaudpawlak
It's more like first case.  There is an application that allows you to write scripts(macros)  in Javascript which are executed within application. This application provides API which you can use in Javascript. For example there is a class 'Session'  which has method 'output'  which writes any text to console. 

I treated that API as external JS lib and wrote candy for it. Works like a charm :-) apart from minor things described above

**Comment by panzupa** [comment]
@renaudpawlak I have a little problem with generating proper candy. I moved my API files to new project.
According to https://github.com/cincheo/jsweet/blob/v1.0.0/doc/jsweet-language-specifications.md#packaging-a-jsweet-jar-candy I try to generate candy.d.ts file

> Typically, d.ts files shall be generated by the JSweet transpiler using the following options:
> declaration: turns on the generation of the d.ts files.
> dtsout: tells where to put the generated d.ts files (for packaging, place them in 
> RESOURCES/src/typings, where RESOURCES is your Maven project resources directory).

So I added to pom.xml

```
<declaration>true</declaration>
<dtsout>src/main/resources/src/typings</dtsout>
```

My maven plugin configuration looks like that:
```
<plugin>
<groupId>org.jsweet</groupId>
				<artifactId>jsweet-maven-plugin</artifactId>
				<version>1.1.1</version>
				<configuration>
					<verbose>true</verbose>
					<tsOut>target/ts</tsOut>
					<outDir>target/js</outDir>
					<targetVersion>ES3</targetVersion>			
					<declaration>true</declaration>
					<dtsout>src/main/resources/src/typings</dtsout>
				</configuration>
				<executions>
					<execution>
						<id>generate-js</id>
						<phase>generate-sources</phase>
						<goals>
							<goal>jsweet</goal>
						</goals>
					</execution>
				</executions>
</plugin>
```

In generated file there is no *.d.ts file.

Could you please help me and say what am I doing wrong?

**Comment by renaudpawlak** [comment]
First, just in case, make sure that you use the sepcs for your version: https://github.com/cincheo/jsweet/blob/v1.1.1/doc/jsweet-language-specifications.md#packaging-a-jsweet-jar-candy

Then, please take example on the J4TS project, with creates and deploy a candy. Here is the [pom.xml](https://github.com/cincheo/j4ts/blob/master/pom.xml).

That said, I don't see any obvious mistake in your ``pom.xml``. So it might be the content of your files that causes trouble. Have you checked that the ``bundle.ts`` file is not empty? Also, what I suspect is that if you have only @Ambient classes in your files, the declarations are empty, because the declarations are already in your ``bundle.ts`` file. If that is the case, I would then copy the ``bundle.ts`` into my ``bundle.d.ts`` file (see the J4TS pom.xml to have an example on how to copy a file after it has been generated).

If I am getting it right, it's an interesting use case... Your are trying to build a candy for an existing JavaScript library using JSweet. What we have been doing so far is to build a candy out of a d.ts file, or build a candy out of a JSweet program. That's probably why it is not well documented, but I hope that your work will help to improve it :)

**Comment by panzupa** [comment]
Thanks for help @renaudpawlak 

Generaly I tried what you wrote but without any success.

> what I suspect is that if you have only @Ambient classes in your files, the declarations are empty

That's correct, my java files look like that

> @Ambient
> public class Session {
> 	public String UserName;
> 	public String Version;
> 
> 	public static native String Input(String Prompt);
> }
All of them, each package has package-info.java file as well.

I think the problem is that js files are empty as well.

Generally all *.js and *.ts files are empty. I triedy different combinations with eclipse plugin and maven. Allways the same.

> If I am getting it right, it's an interesting use case... Your are trying to build a candy for an existing JavaScript library using JSweet. What we have been doing so far is to build a candy out of a d.ts file, or build a candy out of a JSweet program. That's probably why it is not well documented, but I hope that your work will help to improve it :)

I guess, It looks like reverse engineering of a JS lib from your perspective. While I was 
reading docs I was asking myself "what the hell is that d.ts " :-)

Good thing is that output of a program that uses my "candy" is perfectly usable.  However, I'm determined to pack it as a proper Candy. If you're interested I can share with you my eclipse project. Maybe it's matter of machine/configuration?

**Comment by renaudpawlak** [comment]
If it's ok with you, I'd be interested in sharing your project. I am currently working on a quite similar case with Angular 2 and I believe that it could become a very common use case in the future. The idea would be to come up with some project templates. You can use my personal email address for further communication if you wish.
